streams/shared/src/main/scala/zio/stream/StreamCursor.scala

```diff
@@ -0,0 +1,30 @@
+package zio.stream
+
+import zio._
+
+/**
+ * A cursor that allows effectful iteration over a stream. Usage is similar to
+ * [[ZStream#Pull]], but additionally offers the option to traverse a stream multiple
+ * times.
+ */
+trait StreamCursor[-R, +E, +A] {
+
+  /**
+   * Pull the next value of the stream. As with [[ZStream#Pull]], `fail(None)` signals end of stream.
+   * If any cursor has advanced past this point already, the effect will not be reevaluated.
+   */
+  val next: ZIO[R, Option[E], Chunk[A]]
```

Comment:
Would be useful to have a variant that you can return: `ZIO[R, Option[E], A]` like the private `BufferedPull` as this is sometimes more convenient than processing things in chunks, you need a second cursor into the current chunk.

Reply:
I agree that it can be convenient to have, but I feel this is better solved by working with the stream representation.
We could include a `next: ZIO[R, Option[E], A]` method, but that would mean keeping at least one additional ref around per cursor.

That makes me a bit reluctant to include it. Any other opinions?